stan-dev/stan
# Add deserializer class for reading constraints easier

#### Submission Checklist


#### Summary

This does the FR for #3005 by making a `deserializer` class that handles constraints for the types and bounds used in the Stan compiler. It's a bit of a monster but handles pretty much all the parameter sizes we have in Stan. The idea here is that instead of the compiler generating code for paremeters like 

```cpp
      local_scalar_t__ sd_1 = in__.scalar();
      current_statement__ = 5;
      if (jacobian__) {
        current_statement__ = 5;
        sd_1 = stan::math::lb_constrain(sd_1, 0, lp__);
      } else {
        current_statement__ = 5;
        sd_1 = stan::math::lb_constrain(sd_1, 0);
      }
```

It can instead just use

```cpp
      local_scalar_t__ sd_1 = in__.read_lb<local_scalar_t__, Jacobian__>(lp__, 0);
```

This works for eigen matrices and arrays containing arrays so something like the following is perfectly legal

```cpp
      using big_vec = std::vector<std::vector<var_value<Eigen::Matrix<double, -1, -1>>>>;
      big_vec sd_1 = in__.read_lb<big_vec, Jacobian__>(0, lp__, 3, 3, 3, 3);
```

For general reading the function is `read<RETURN_TYPE>(dims)` where `RETURN_TYPE` is the type requested back and dims are the dimensions of the type. All of the constrain functions have the form

```cpp
read_{constrain}<{return_type}, Jacobian>({constrain_args}, lp, {dims})
```

Where 
- `{constrain}` is the kind of the constraint like `lb,ub,lub` etc, 
- `{return_type}` is the requested return type, 
- `Jacobian` is whether to account for the Jacobian calculation
- {constrain_args} is the arguments passed to the actual constrain function
-  lp is the the log prob that will be incremented if `Jacobian` is true
-  {dims} are the dimensions of the return type

This also adds `read()` methods for complex scalars, vectors, and matrices.

#### Intended Effect

This should simplify some of the compiler implementation for things like the new matrix type while also making things more performant by utilizing vectorized versions of the constrain functions when they are available.

#### How to Verify

About 2/3rds of this PR are testing, where we run all the standard tests from `reader`, then also check `std::vector` return types and `var_value<Eigen>` returns for all matrix tests. You can run the tests with

```
./runTests.py ./src/test/unit/io/deserializer_test.cpp \
 ./src/test/unit/io/deserializer_varmat_test.cpp \
 ./src/test/unit/io/deserializer_stdvector_test.cpp
```

#### Side Effects

There's two things I still need to add here, tests for the complex stuff and tests for matrices with the `offset_and_multiplier` constraints. I may need to make a Stan math PR for offset multiplier 

#### Documentation

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Steve Bronder



By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
